Swap to `gem_layout`

alex-ju opened the pull request:
### What

Update `whitehall` frontend to use the Design System-based page layout

### Why

Benefits include performance improvements, accessibility improvements, and consistency with other applications.

### Visual Changes
No unexpected visual changes. There are a few instances where, for example, headings were missing styles

When reviewing, I recommend deploying in integration for a visual check. I've added a set of pages that can be used for this.

- https://www.gov.uk/government/how-government-works
- https://www.gov.uk/government/ministers
- https://www.gov.uk/government/history
- https://www.gov.uk/government/history/past-prime-ministers
- https://www.gov.uk/government/history/past-prime-ministers/harold-wilson
- https://www.gov.uk/government/history/past-chancellors
- https://www.gov.uk/government/history/past-foreign-secretaries
- https://www.gov.uk/government/history/past-foreign-secretaries/charles-fox
- https://www.gov.uk/government/topical-events/the-uks-industrial-strategy
- https://www.gov.uk/courts-tribunals
- https://www.gov.uk/government/fields-of-operation
- https://www.gov.uk/government/fields-of-operation/iraq
- https://www.gov.uk/world
- https://www.gov.uk/world/organisations
- https://www.gov.uk/world/organisations/british-antarctic-territory
- https://www.gov.uk/world/embassies
- https://www.gov.uk/world/albania/news
- https://www.gov.uk/world/uk-mission-to-the-wto-un-and-other-international-organisations-geneva
- https://www.gov.uk/government/latest?world_locations%5B%5D=albania
- https://www.gov.uk/government/organisations/charity-Commission/about
- https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/893819/exrates-monthly-0720.csv/preview


### Checklist

- [x] Fix whitehall wrappers and layout issues
- [x] Use `govuk-frontend` classes to style basic elements 
- [x] Replace `govuk_frontend_toolkit` font mixins with `govuk-frontend` equivalents (this work is required as `govuk_frontend_toolkit` mixins are referencing the nta font, which is made available by `govuk_template`)

[Trello card](https://trello.com/c/KfRfAxD6/882-update-whitehall-app-to-use-the-public-layout-component)

alex-ju commented:
Rebased to master – ready for review.

maxgds commented on app/views/admin/worldwide_offices/index.html.erb:
```diff
@@ -11,8 +11,8 @@
         <ul class="nav nav-tabs">
-          <li class="active"><a href="#manage-all" data-toggle="tab">All</a></li>
-          <li><a href="#on-home-page" data-toggle="tab">Order on home page</a></li>
+          <li class="active"><a class="govuk-link" href="#manage-all" data-toggle="tab">All</a></li>
```
This looks like an admin template, I didn't think we were applying the layout there, should this be changing?

alex-ju replied:
Good spot! Indeed an admin template, will revert the change in this file.

maxgds commented:
Corporate information pages have mostly moved out of Whitehall, but there's [a ticket](https://trello.com/c/f1W1xQ4R/70-cleanup-duplicate-corporate-information-pages-for-charity-commission) to fix a problem whereby some are duplicates and left behind. Meanwhile, though, as you've updated the template it looks like it needs a little more attention. Specifically I've noted the contents of the left hand column have grown and moved into the main column, and the tiny text in the middle doesn't have the right colour applied to the links. On the plus side, the font sizes are more consistent!

https://www.gov.uk/government/organisations/charity-Commission/about 
https://www.integration.publishing.service.gov.uk/government/organisations/charity-Commission/about

alex-ju commented:
> Corporate information pages have mostly moved out of Whitehall, but there's [a ticket](https://trello.com/c/f1W1xQ4R/70-cleanup-duplicate-corporate-information-pages-for-charity-commission) to fix a problem whereby some are duplicates and left behind. Meanwhile, though, as you've updated the template it looks like it needs a little more attention. Specifically I've noted the contents of the left hand column have grown and moved into the main column, and the tiny text in the middle doesn't have the right colour applied to the links. On the plus side, the font sizes are more consistent!
> 
> https://www.gov.uk/government/organisations/charity-Commission/about https://www.integration.publishing.service.gov.uk/government/organisations/charity-Commission/about

The charity commission seems to be a good page to test this change with – thanks for pointing it out! I was only able to identify a few that were pretty much laking content. Will address the layout issue and try to get it close to its current design.

maxgds commented:
The "keyline header" class on pages like this one looks quite overbearing with the new font - it's a bit much with the bold text and the bold underline. Should we maybe just kill the class and standardise the headers to be as per design system standard? I'm uncertain as I know we're trying to switch out the essentials to just put the layout in and that's maybe a step further but it looks bad to me. 

![image](https://user-images.githubusercontent.com/31649453/141811169-dad37c51-863f-4d18-8a99-cbce8df0f36d.png)
https://www.integration.publishing.service.gov.uk/world/organisations/british-antarctic-territory

maxgds commented on app/views/documents/_filter_selections.mustache:
```diff
@@ -1,48 +1,48 @@
 <span class="count">{{result_count}}</span> <strong>{{pluralized_result_type}}</strong>
```
I'm pretty sure this is another admin only piece of code that should be reverted, based on previous PRs against it, eg https://github.com/alphagov/whitehall/pull/4334

alex-ju replied:
Pagination for the admin area, indeed. Have reverted this.

maxgds commented:
Some missing govuk-link classes here, it seems:
https://www.integration.publishing.service.gov.uk/government/latest?world_locations%5B%5D=albania

maxgds commented:
Latest articles on this page have gained a lot of extra padding, probably as a result of applying govuk-list:
https://www.integration.publishing.service.gov.uk/world/uk-mission-to-the-wto-un-and-other-international-organisations-geneva

maxgds commented on app/views/kaminari/_next_page.html.erb:
```diff
@@ -7,4 +7,4 @@
     remote:        data-remote
```
Can't find any use of these kaminari things - do you have examples? Maybe they're only admin side?

alex-ju replied:
Looked into this, and it seems like kaminari is only used for pagination in the admin area. Have reverted these changes as well.

maxgds commented on app/assets/stylesheets/frontend/helpers/_worldwide-organisation-header.scss:
```diff
@@ -21,15 +21,18 @@
-      margin: 0;
+      @include govuk-text-colour;
+      @include govuk-font($size: 16);
+      @include govuk-responsive-margin(4, "bottom");
```
I'm not sure but I think this may be over-riding the margin set to 0 below. 

We've gained extra whitespace at the top of this page:
https://www.integration.publishing.service.gov.uk/world/organisations/british-consulate-izmir
It's caused by extra padding at the bottom of the dl originally set to 0 on a selector of #whitehall-wrapper .worldwide-organisation-header .metadata dl

alex-ju replied:
Fixed this by [removing the page header padding](https://github.com/alphagov/whitehall/pull/6240/commits/5286f78753eea388d0a0554e33748d7d0b9dba84). I prefer <dl> elements to have a consistent margin bottom and keep the spacing in only one direction as much as possible.

maxgds commented on app/views/contacts/_organisation_contact.html.erb:
```diff
@@ -18,21 +18,21 @@
       <div class="govuk-grid-column-one-third">
         <% if contact.email.present? %>
-          <p class="govuk-!-margin-bottom-4">
+          <p class="govuk-body govuk-!-margin-bottom-4">
```
We need the govuk-body class applied to the address tag on line 13 as well, currently the font is wrong:https://www.integration.publishing.service.gov.uk/world/organisations/british-antarctic-territory

alex-ju commented:
> That was a mammoth checking/testing session! I hope you don't mind that I drip fed the suggestions/comments through rather than collecting them for a single review. I think I'm done now, though. The nature of Whitehall and these changes make me think things may still slip through but for the most part we're good and we can address what we spot on an ad hoc basis later.

Thanks so much for taking the time to review this properly! It sounds like a few things slipped and I'll do my best to address your feedback in the next few days to keep momentum on this.

alex-ju commented:
I tried to address all the comments above. There are still two areas where we'll have visual changes, as you mentioned.

> Corporate information pages have mostly moved out of Whitehall, but there's [a ticket](https://trello.com/c/f1W1xQ4R/70-cleanup-duplicate-corporate-information-pages-for-charity-commission) to fix a problem whereby some are duplicates and left behind. Meanwhile, though, as you've updated the template it looks like it needs a little more attention. Specifically I've noted the contents of the left hand column have grown and moved into the main column, and the tiny text in the middle doesn't have the right colour applied to the links. On the plus side, the font sizes are more consistent!
> 
> https://www.gov.uk/government/organisations/charity-Commission/about https://www.integration.publishing.service.gov.uk/government/organisations/charity-Commission/about

Thanks for pointing out the card related to this. It sounds like these pages are potentially going to live only temporarily in whitehall then. I fixed the issues with the layout on the page and used standard heading styles (open to suggestions for other approaches here, but the old headings look like nothing across the whole website from my perspective).

<table>
<tr><th>Before</th><th>After</th></tr>
<tr><td valign="top">

![corporate-info-before](https://user-images.githubusercontent.com/788096/142617264-d9a23d54-7382-42f7-a133-35a79b98dd2c.png)

</td><td valign="top">

![corporate-info-after](https://user-images.githubusercontent.com/788096/142617273-9a6d377e-7ec6-4afb-a5db-b37265edd51f.png)

</td></tr>
</table>

------

> The "keyline header" class on pages like this one looks quite overbearing with the new font - it's a bit much with the bold text and the bold underline. Should we maybe just kill the class and standardise the headers to be as per design system standard? I'm uncertain as I know we're trying to switch out the essentials to just put the layout in and that's maybe a step further but it looks bad to me.
> 
> ![image](https://user-images.githubusercontent.com/31649453/141811169-dad37c51-863f-4d18-8a99-cbce8df0f36d.png) https://www.integration.publishing.service.gov.uk/world/organisations/british-antarctic-territory

I like the suggestion of dropping the "keyline header" classes so I relied on the same styling for this h2 element as the other two on the page.

<table>
<tr><th>Before</th><th>After</th></tr>
<tr><td valign="top">

![world-organisation-before](https://user-images.githubusercontent.com/788096/142617295-e4ba7362-ac18-4ba8-96e0-591a9c1a47c6.png)

</td><td valign="top">

![world-organisation-after](https://user-images.githubusercontent.com/788096/142617313-b7edb8b4-53b6-4ab1-90ab-747029ff154e.png)

</td></tr>
</table>

maxgds commented:
I deployed the branch to integration yesterday and had another look around and I've not spotted anything further. I also had a cursory poke around the admin side to satisfy myself that nothing drastic had accidentally crept through into there and didn't spot anything. I think we're good to ship this (and look out for reports of anything we missed, of course!) I would re-review and approve but I'm aware there's some extra work needed to deal with some of the assets for the Feedback component, so I'll await that addition first.

alex-ju commented:
> I deployed the branch to integration yesterday and had another look around and I've not spotted anything further. I also had a cursory poke around the admin side to satisfy myself that nothing drastic had accidentally crept through into there and didn't spot anything. I think we're good to ship this (and look out for reports of anything we missed, of course!) I would re-review and approve but I'm aware there's some extra work needed to deal with some of the assets for the Feedback component, so I'll await that addition first.

Thanks, Max! I also did manual checks on the admin area, just to be on the safe side. Have just rebased to `main`, squashed a few commits to have a clean git history and re-deployed the branch in integration (in case there's anything else you'd like to check).

In regards to the feedback component, I tried removing the assets at the application level, but unfortunately it's not as simple as it should be. The redundant styling needs to remain in place until a proper fix is set in place at the component level. The current implementation relies on class composition `govuk-button gem-c-feedback__prompt-link` and if the `feedback.scss` is defined in `static`, then a follow-up import of `button.scss` at the application level will take precedence, resulting in the following visual. This is a classic example of where class composition is not working to our advantage, as it compromises the single responsibility promise of BEM.

<img width="960" alt="Screenshot 2021-11-24 at 13 14 37" src="https://user-images.githubusercontent.com/788096/143247350-03596488-5153-4159-a2bc-3265c7441363.png">

Considering this, I would suggest addressing the feedback asset redundancy and styling in a separate PR as we have to adjust the component first anyway and address the issues across the frontend application.

maxgds reviewed (approved):
We've been over-written on integration by another deploy but I'm happy to sign this off as good now.
